Add dropdown task default annotations on update

mcbouslog opened the pull request:
Staging branch URL: https://dropdown-default-update-fix.pfe-preview.zooniverse.org/projects/markb-panoptes/test-project-1-mb/classify?reload=1&workflow=3165

Closes #4401. 

Describe your changes.
- check and if annotations not present, set default annotations (`{ option: false, value: null }`) on component update

# Required Manual Testing

- [x] Does the non-logged in home page render correctly?
- [x] Does the logged in home page render correctly?
- [x] Does the projects page render correctly?
- [x] Can you load project home pages?
- [x] Can you load the classification page?
- [x] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [x] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

eatyourgreens commented:
Should there be tests for these defaults in the dropdown task tests?

eatyourgreens commented:
Not a blocking comment but the naming here is a little bit confusing. The default annotation for a dropdown task is an empty array. Is this setting default initial values for the dropdown menus, then adding those to the empty annotation? If so, it might need to call the `onChange` handler to register the new values in classifier state.

mcbouslog commented:
Great point @eatyourgreens! I have a meeting this morning, but would then like to check and edit accordingly, please hold on merging.

eatyourgreens commented:
I'm not 100% sure about needing to call `onChange`, since the current component works without doing that. That might be because the current code changes `props.annotation.value` which then changes `state.annotations` in the classifier, indirectly.

mcbouslog commented:
Added tests.

@eatyourgreens I think your most recent comment is correct, it was working due to changing props. I refactored so on mount and on update it calls `setDefaultAnnotations` (also don't like name, open to change) which iterates through selects with their index and calls `onChangeSelect`. Which I think is preferable? Let me know what you think, happy to change back.

eatyourgreens commented:
I was thinking something like this to set up the initial values. I'm not sure if my coffeescript syntax is correct:

```coffeescript
value = @props.task.selects.map () -> { option: false, value: null }
newAnnotation = Object.assign {}, @props.annotation, { value }
@props.onChange newAnnotation
```

I'm not sure if it's better to call `onChange` once, with the new annotation, or once for each select. It looks like your current code works, so this isn't a blocking issue.

mcbouslog commented:
That's more elegant, only makes one `onChange` call - updated accordingly.

I think this is ready then, I'll assign to you to merge when comfortable with the recent change, thanks Jim!

eatyourgreens commented:
👍 Looks good to me. Calling `onChange` immediately on mount seems like an odd pattern, just to set some initial values, but I think fixing that would involve changes to how all the task components work, and is outside the scope of this PR.

eatyourgreens commented on app/classifier/tasks/dropdown/index.cjsx:
```diff
@@ -90,16 +90,22 @@ module.exports = createReactClass
+  setDefaultValues: ->
+    unless @props.annotation.value.length
+      value = @props.task.selects.map () -> { option: false, value: null }
+      newAnnotation = Object.assign @props.annotation, {value}
```
There's a missing {} here. This modifies `@props.annotation`, rather than copying it and creating a new annotation.

`newAnnotation = Object.assign {}, @props.annotation, {value}

mcbouslog commented:
Yes, makes sense. And I think I should change [the similar line in `onChangeSelect`](https://github.com/zooniverse/Panoptes-Front-End/blob/master/app/classifier/tasks/dropdown/index.cjsx#L205), to the same:
```
newAnnotation = Object.assign {}, @props.annotation, {value}
```

eatyourgreens commented:
Good idea. My reason for doing this is because `props.annotation` might be a reference to the original variable in the Classifier component.

mcbouslog commented:
Yep makes sense, good and important refactoring.
